File: frontend/playground/battle/collision/detector/box-surface-collisions-detector.ts
import {Collision, type CollisionsDetector} from "~/playground/battle/collision/collision";
import {BattleCalculations, BoxCalculations, type Calculations} from "~/playground/data/calculations";
import {SurfaceContactUtils} from "~/playground/utils/surface-contact-utils";
import {BodyUtils} from "~/playground/utils/body-utils";

export class BoxSurfaceCollisionsDetector implements CollisionsDetector {
  detect(calculations: Calculations, battle: BattleCalculations): Set<Collision> {
    if (calculations instanceof BoxCalculations) {
      return this.detectBoxCollisions(calculations, battle)
    }
    return new Set()
  }

  private detectBoxCollisions(box: BoxCalculations, battle: BattleCalculations): Set<Collision> {
    const collisions = new Set<Collision>()
    const bodyPart = BodyUtils.getBodyPart(box.model.specs.shape, box.getGeometryNextPosition()!)
    const contacts = SurfaceContactUtils.getContacts(bodyPart, battle.model.room, true)
    for (const contact of contacts) {
      collisions.add(Collision.withSurface(box, contact))
    }
    return collisions
  }
}

File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/collision/detector/box/BoxSurfaceCollisionsDetector.java
package com.github.aadvorak.artilleryonline.battle.collision.detector.box;

import com.github.aadvorak.artilleryonline.battle.calculations.BattleCalculations;
import com.github.aadvorak.artilleryonline.battle.calculations.BoxCalculations;
import com.github.aadvorak.artilleryonline.battle.calculations.Calculations;
import com.github.aadvorak.artilleryonline.battle.collision.Collision;
import com.github.aadvorak.artilleryonline.battle.collision.detector.CollisionsDetector;
import com.github.aadvorak.artilleryonline.battle.common.lines.BodyPart;
import com.github.aadvorak.artilleryonline.battle.utils.SurfaceContactUtils;
import org.springframework.stereotype.Component;

import java.util.Set;

@Component
public class BoxSurfaceCollisionsDetector implements CollisionsDetector {

    @Override
    public Set<Collision> detect(Calculations<?> calculations, BattleCalculations battle, boolean first) {
        if (calculations instanceof BoxCalculations boxCalculations) {
            return detect(boxCalculations, battle);
        }
        return Set.of();
    }

    private Set<Collision> detect(BoxCalculations box, BattleCalculations battle) {
        var bodyPart = BodyPart.of(box.getGeometryNextPosition(), box.getModel().getSpecs().getShape());
        return SurfaceContactUtils.getContacts(bodyPart,
                        battle.getModel().getRoom(), true).stream()
                .map(contact -> Collision.withSurface(box, contact))
                .collect(java.util.stream.Collectors.toSet());
    }
}
